File: python/basics/stack/20_valid_parentheses.py
'''
https://leetcode.com/problems/valid-parentheses/description/

use of stack
map of closing to opening bracket
when you encounter a closing bracket - pop up from stack and the bracket should match.  If no or no matching, invalid
    at the end the stack should be empty
'''
def valid_parentheses(input:str):
    bracketmap = {")":"(", "]":"[", "}":"{"}
    result = False

    stk = []
    for c in input:
        if c not in bracketmap.keys():
            stk.append(c)
        else:
            if len(stk) == 0:
                return False
            pop = stk.pop()
            if pop != bracketmap[c]:
                return False
    return len(stk) == 0


input = "()"
result = valid_parentheses(input)
print("result: " + str(result))
#true


input = "()[]{}"
result = valid_parentheses(input)
print("result: " + str(result))

#true

input = "(]"
result = valid_parentheses(input)
print("result: " + str(result))
#false

input = "([])"
result = valid_parentheses(input)
print("result: " + str(result))

#true

File: problem_solving/stack/cpp/20_valid_parentheses.cpp
/*
https://leetcode.com/problems/valid-parentheses/

iterate the string
when closing brace is encountered, check if the top of stack has the correcsponding opening, if yes, pop, if no, return false.
for opening brackets, push the stack.
in the end the stack should be empty for success.
*/
#include<stack>
#include<unordered_map>
#include<iostream>

bool valid_parentheses(const std::string& s) {
    std::unordered_map<char, char> braces = {
        {')', '('},
        {']', '['},
        {'}', '{'}
    };

    std::stack<char> st;
    for( const auto ch : s ) {
        //end brace found...
        if(braces.find(ch) != braces.end() ) {
            if(st.empty()) {
                return false;
            }
            auto top = st.top();
            if(braces[ch] != top ) {
                return false;
            }
            st.pop(); 
        } else {
            st.push(ch);
        }
    }
    if (st.empty()) {
        return true;
    }
    return false;
}


int main() {
    std::string s = "()";
    bool result = valid_parentheses(s);
    std::cout << "result: " << result << std::endl;
    //Output: true


    s = "()[]{}";
    result = valid_parentheses(s);
    std::cout << "result: " << result << std::endl;
    //Output: true

    s = "(]";
    result = valid_parentheses(s);
    std::cout << "result: " << result << std::endl;
    //Output: false

    s = "([])";
    result = valid_parentheses(s);
    std::cout << "result: " << result << std::endl;
    //true

    return 0;
}